R/nm-join.R

```diff
@@ -0,0 +1,179 @@
+#' the joined data. If `getOption(bbr.verbose) == TRUE` a message will be
+#' printed about any columns dropped this way.
+#'
+#' The one exception to this is the `DV` column. If `DV` is present in the input
+#' data _and_ at least one of the table files, the `DV` column from the input
+#' data will be renamed to `DV.DATA` and the column from the table file kept as
+#' `DV`.
+#'
+#' @importFrom dplyr left_join select
+#' @seealso [nm_tables()], [nm_table_files()], [nm_file()]
+#' @export
+nm_join <- function(
+  .mod,
+  .join_col = "NUM",
+  .files = nm_table_files(.mod),
+  .superset = FALSE,
+  .bbi_args = list(
+    no_grd_file = TRUE,
+    no_shk_file = TRUE
+  )
+) {
+
+  if (inherits(.mod, "character")) {
+    checkmate::assert_string(.mod)
+    .mod <- read_model(.mod)
+  }
+  check_model_object(.mod, c(NM_MOD_CLASS, NM_SUM_CLASS))
+  checkmate::assert_string(.join_col)
+  checkmate::assert_character(.files)
+  checkmate::assert_logical(.superset, len = 1)
+  checkmate::assert_list(.bbi_args)
+
+  df_list <- nm_tables(.mod, .files = .files)
+  .d <- df_list$data
+  .tbls <- df_list[2:length(df_list)]
+  if (
+    "DV" %in% names(.d) &&
+    "DV" %in% unlist(map(.tbls, names))
+  ) {
+    .d <- rename(.d, DV.DATA = .data$DV)
+  }
+  col_order <- names(.d)
+
+  .join_col <- toupper(.join_col)
+  if (!(.join_col %in% names(.d))) {
+    stop(glue("couldn't find `.join_col` {.join_col} in data with cols: {paste(names(.d), collapse = ', ')}"))
+  }
+
+  if (.superset) {
+    join_fun <- function(x, y, ...) left_join(y, x, ...)
+  } else {
+    join_fun <- left_join
+  }
+
+  # get number of ID's and records
+  .s <- if (!inherits(.mod, NM_SUM_CLASS)) {
+    model_summary(.mod, .bbi_args = .bbi_args)
+  } else {
+    .mod
+  }
+  nid <-  .s$run_details$number_of_subjects
+  nrec <- .s$run_details$number_of_data_records
+
+  # do the join(s)
+  for (.n in names(.tbls)) {
+    tab <- .tbls[[.n]]
+    has_id <- "ID" %in% names(tab)
+
+    # skip table if nrow doesn't match number of records or ID's
+    # because if neither is true than this is the wrong kind of file
+    # (or something is wrong with NONMEM output)
+    if (!(nrow(tab) %in% c(nrec, nid))) {
+      warning(glue("{.n} skipped because number of rows ({nrow(tab)}) doesn't match number of records ({nrec}) or IDs ({nid})"), call. = FALSE)
+    }
+
+    # if $FIRSTONLY table join on ID
+    if (nrow(tab) == nid) {
+      verbose_msg(glue("{.n} is $FIRSTONLY table"))
+
+      # if ID is missing, get it from the data by using .join_col
+      if (!has_id) {
+        tab <- tab %>%
+          left_join(select(.d, .data$ID, !!.join_col), by = .join_col)
+      }
+
+      # toss .join_col, if present, because we're joining on ID
+      tab[[.join_col]] <- NULL
+
+      # do the join
+      tab <- drop_dups(tab, .d, "ID", .n)
+      col_order <- union(col_order, names(tab))
+      .d <- join_fun(tab, .d, by = "ID")
+    }
+
+    if (nrow(tab) == nrec) {
```

Comment:
The `== nid` and `== nrec` should never both be followed, right?  In that case, I think it'd be clearer to restructure the three outer `if`s above into `if ... else if ... else ...` to make invalid states inaccessible.

Reply:
I can never decide what is the least annoying way to do this in R. It's basically a case statement, but those are hideous in R. I was kinda using the "left justified path" mentality, but I think you're right that if/else is probably cleaner here.